### PR TITLE
docs/tutorial.md: in the rebase example, use commit descriptions rather than change ids

### PR DESCRIPTION
**Pull request by gasche**
The tutorial currently says:

> Let's now rebase B2 directly onto A:
>
>     $ jj rebase -s puqltuttrvzp -d nuvyytnqlquo

I find it fairly confusing for the following reasons:

- The change ids are different on my machine, so it is difficult for me to tell which of puqltuttrvzp or nuvyytnqlquo correspond to A or B2. I have to scroll above in the tutorial to locate this information again, and I don't want to have to do this.

- You haven't explained the meaning of `-s` and `-d` yet, so I cannot guess from those either.

Instead I propose to select the commits from their commit message / description:

    $ jj rebase -s 'description(B2)' -d 'description(A)'

This is a tad verbose but much clearer. As a reader of the tutorial I appreciate clarity. (I just added both `descr(x)` and `msg(x)` as shorter alises of `description(x)` to my config file, but obviously the tutorial cannot rely on those.)

### Extra comments

1. I added a mention of `description(x)` earlier in the tutoriel when revsets are described.

2. In that part I added a link to the full documentation page on revsets. Having this link would have helped me when I decided that I wanted to name the commit by its commit message, but did not know how to.

2bis. I also tried `jj help revsets`, by analogy with `git help revisions`, but this does not work.

3. Using `--source` and `--destination` in the `rebase` line would have made my life easier, as I would not have had to guess the meaning of `-s` and `-d`. Have you considered using full names in the tutorial for each new option, and then the short name on followup uses? Should I consider a PR doing this?

### CODE REVIEW

**Comment by gasche** [comment]
This is my first contribution to jj. I am familiar with CLAs in general, but I find the imposed process here disturbing for two reasons:

1. To contribute to `jj`, I am asked to sign a CLA that applies to *all* Google products. This feels excessive to me, especially since the wording of the agreement basically suggest that all interactions on Google development spaces (not just specifically submitting a code change) are covered by the CLA *by default* unless I *explicitly* mention that a particular remark of mine is not meant to be covered.

2. Before I realized (1), I assumed that after clicking 'I agree' on https://cla.developers.google.com/ , I would be asked which project is covered by the CLA. So I clicked. Then I realized my misconception, and I wanted to un-click. And now I am finding out that there is *no* interface to revoke this agreement. I understand that it is important for the CLA purposes that I grant irrevocable rights to my past contributions. But I should be able to stop granting these rights to Google for future contributions whenever I want, and the process to do this should be easy.

Honestly this makes me very uncomfortable with the idea of contributing to jj again. For something that is "not a Google product" (quoting from the webpage), you are in a situation where a legal process is imposed to you that seems to be very badly designed, in a way that discourages contributions.

**Comment by martinvonz** [comment]
My understanding is that the CLA is to protect the project from accepting code from someone who later decides that the project is not allowed to use their contribution. It does not transfer ownership of the code to Google (not sure if that's your concern).

I googled around a bit without finding any criticism. I also search HN (my impression is that people don't hesitate to complain there) and didn't find anything there either about Google's CLA in particular. I found https://drewdevault.com/2018/10/05/Dont-sign-a-CLA.html, however, but the concern expressed there is that CLAs allow an organization to re-license the code. I don't think that's the case with Google's CLA (but IANAL). See https://google.github.io/opencasebook/cla-policy/#cla for Google's view on the matter.

**Comment by martinvonz** [comment]
And thanks for your contributions, btw! Before I look at the commits, please let me know if you're willing to have it accepted under the CLA. If not, I think it's best that I don't even look at it. My impression is that you're okay with getting this PR in (otherwise you would presumably have closed it) and you're just hesitant about future contributions, but I want to make sure.

**Comment by gasche** [comment]
I have no concern with signing a CLA covering *my contributions to jj*, including one that would let you change the license in the future.

I however have two concern with the current process:

1. I am forced to sign a CLA that applies to *all Google products*, not just jj specifically.
2. The interface to do so does *not* offer the option to revoke my agreement for my future contributions.

So to contribute to your project I have to sign a very broad agreement, and I am not offered the option to change my mind. I find this problematic.

I understand that I am consuming your maintainer time/attention for a not-fun concern on a very small contribution. Apologies! Maybe this discussion should be posted to HN/whatever to understand if other peoples consider this process reasonable. But even if they did, I don't know that you could do to improve it -- I am assuming that you have no choice than to follow the standard Google CLA procedure, and that changing any of it is infeasible in practice.

**Comment by martinvonz** [comment]
> I however have two concern with the current process:
> 
> 1. I am forced to sign a CLA that applies to _all Google products_, not just jj specifically.
> 2. The interface to do so does _not_ offer the option to revoke my agreement for my future contributions.
> 
> So to contribute to your project I have to sign a very broad agreement, and I am not offered the option to change my mind. I find this problematic.

That's fair. Maybe one with complication with (1) is that your contribution might also end up in other project (though that seems very unlikely for a documentation update like this PR). I've asked in our internal chat room for open source questions.

> I understand that I am consuming your maintainer time/attention for a not-fun concern on a very small contribution. Apologies!

No problem! CLA concerns come up once in a while, so figuring out if we can improve anything might benefit the project in the long run.

**Comment by gasche** [comment]
Thanks! For the time being I did sign the CLA so there should be no more issues with that.

File: docs/tutorial.md [hunk 1]
```diff
@@ -163,7 +163,10 @@ There are also operators for getting the parents (`foo-`), children (`foo+`),
 ancestors (`::foo`), descendants (`foo::`), DAG range (`foo::bar`, like
 `git log --ancestry-path`), range (`foo..bar`, same as Git's). There are also a
 few more functions, such as `heads(<set>)`, which filters out revisions in the
-input set if they're ancestors of other revisions in the set.
+input set if they're ancestors of other revisions in the set, and
+`description(foo)`, the commits whose description (commit message) contains `foo`.
```
Comment (martinvonz) [inline docs/tutorial.md:167]: I think we can remove the mention of `heads()` here. It's mentioned here for historical reasons (from when `jj log`'s default revset was too large and we told the user to define an alias using `heads()` here).

File: docs/tutorial.md
Comment (martinvonz) [inline docs/tutorial.md:1]: > (Note: using `--source` and `--destination`
  would make it easy for me to guess which is which, but the tutorial
  consistently makes the choice to use short options for now.)

Yes, I think that would be an improvement.

**Comment by martinvonz** [comment]
I talked a bit to the open-source folks at Google about the CLA issue. I don't think there's much interest in allowing  per-project CLAs. I think it would just make the process more complicated for most people, and it would of course require some development work on the system that keeps track of CLAs. So for now, I'm afraid the best advice I can give is to look out for the "cla/google" CI check when you contribute to a new project and you're not sure if your contribution is going to be covered by the CLA you signed.

File: docs/tutorial.md
Comment (gasche) [reply docs/tutorial.md:1]: I can volunteer to edit the tutorial to consistently use the long name of each option first, if there is interest. I would propose to agree on a consistent wording/formulation first.

Currently the tutorial uses, for example:

> We can use the `-r` flag to select a different set of revisions to list. [...]
>
>     $ jj log -r '@ | root() | branches()'

If we follow the principle of using the long name on the first use of each option, what phrasing would you like?

I would propose the following:

> We can use the `--revisions` flag (`-r`) to select a different set of revisions to list. [...]
>
>     $ jj log --revisions '@ | root() | branches()'   # short: jj log -r ...

Principles;

1. In text, mention the short name in parentheses.
2. In the command, mention the short name in a comment (with arguments elided).
3. Other uses of the same option in the tutorial remain unchanged.
 
(Note in passing: I find it disturbing that `jj foo -h` and `jj foo --help` do not have the same meaning, with `-h` showing only a help summary. I wasted some time looking for some information that was missing from `jj config -h`, without thinking of asking for `jj config --help` instead.)

File: docs/tutorial.md (hunk 1 above)
Comment (gasche) [reply docs/tutorial.md]: I pushed a new commit to remove the mention of `heads(x)`.

File: docs/tutorial.md
Comment (gasche) [reply docs/tutorial.md:1]: (I think this should rather be done in a separate PR.)

**Comment by gasche** [comment]
After using `jj` just slightly more, my impression is that referring to commits by parts of their descriptions is not necessarily a good approach in real situations. Indeed, a mention of a word or two is likely to be ambiguous and refer to several commits, making commands fail (or have unexpected behavior). So this is not necessarily something that tutorial readers would use after they completed the tutorial. I still believe that it is better than using the Change ID at this point in the tutorial, to avoid the issue that the change IDs on my machine are different from those in the tutorial.

File: docs/tutorial.md
Comment (martinvonz) [reply docs/tutorial.md:1]: > Principles;
> 
> 1. In text, mention the short name in parentheses.
> 2. In the command, mention the short name in a comment (with arguments elided).
> 3. Other uses of the same option in the tutorial remain unchanged.

Another option is to include both long and short names in text and always use the short name in the command, like this:

> We can use the `--revisions`/`-r` flag to select a different set of revisions to list. [...]
> ```
> $ jj log -r '@ | root() | branches()'
> ```

That makes the command line match how I think most users will type it.

What were you thinking of doing for the `jj rebase -s X -d Y` case? We don't currently explain the flags in text there. We could say something like this:

> [...] Let's now rebase B2 directly onto A by using the `jj rebase` command with the `--source`/`-s` and `--destination`/`-d` flags:
> ```
> $ jj rebase -s puqltuttrvzp -d nuvyytnqlquo
> ```

There might be better ways of saying that.

> (Note in passing: I find it disturbing that `jj foo -h` and `jj foo --help` do not have the same meaning, with `-h` showing only a help summary. I wasted some time looking for some information that was missing from `jj config -h`, without thinking of asking for `jj config --help` instead.)

I agree! We get that behavior from clap. There's some discussion in https://github.com/clap-rs/clap/issues/1015.

**Comment by martinvonz** [comment]
> After using `jj` just slightly more, my impression is that referring to commits by parts of their descriptions is not necessarily a good approach in real situations. Indeed, a mention of a word or two is likely to be ambiguous and refer to several commits, making commands fail (or have unexpected behavior). So this is not necessarily something that tutorial readers would use after they completed the tutorial. I still believe that it is better than using the Change ID at this point in the tutorial, to avoid the issue that the change IDs on my machine are different from those in the tutorial.

I think that makes sense. We could include a note like "You will probably mostly refer to commits by change ID in practice, but we use `description()` here to avoid depending on a specific change ID.". What do you think?

**Comment by necauqua** [comment]
Another option is to use the shortest changeid prefix, which for an initial repo is one-two characters long, and it also shows users that this is a thing - I mostly use them, for example

And they are highlighted in the default log template too
image

File: docs/tutorial.md
Comment (gasche) [reply docs/tutorial.md:1]: I introduce all options consistently using the `<long> / <short>` style in #2415. I consider it an alternative to the prsent PR, which I think could be closed if we prefer to go this route.

**Comment by gasche** [comment]
Closing in favor of #2415.

**Pull request closed without merging**